lib/service/awsoidc.go

```diff
@@ -0,0 +1,263 @@
+	"github.com/gravitational/teleport/api/utils/retryutils"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+	"github.com/gravitational/teleport/lib/integrations/awsoidc"
+	"github.com/gravitational/teleport/lib/utils/interval"
+)
+
+const (
+	// updateDeployAgentsInterval specifies how frequently to check for available updates.
+	updateDeployAgentsInterval = time.Minute * 30
+
+	// updateDeployAgentsRateLimit specifies the time between updates across AWS regions.
+	updateDeployAgentsRateLimit = time.Second * 30
+)
+
+func (process *TeleportProcess) periodUpdateDeployServiceAgents() error {
+	if !process.Config.Proxy.Enabled {
+		return nil
+	}
+
+	// start process only after teleport process has started
+	if _, err := process.WaitForEvent(process.GracefulExitContext(), TeleportReadyEvent); err != nil {
+		return nil
+	}
+	process.log.Infof("The new service has started successfully. Checking for deploy service updates every %v.", updateDeployAgentsInterval)
+
+	resp, err := process.getInstanceClient().Ping(process.GracefulExitContext())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !resp.ServerFeatures.AutomaticUpgrades {
+		return nil
+	}
+
+	periodic := interval.New(interval.Config{
+		Duration: updateDeployAgentsInterval,
+		Jitter:   retryutils.NewSeventhJitter(),
+	})
+	defer periodic.Stop()
+
+	for {
+		if err := process.updateDeployServiceAgents(process.GracefulExitContext(), process.getInstanceClient()); err != nil {
+			process.log.Warningf("Update failed: %v. Retrying in ~%v", err, updateDeployAgentsInterval)
+		}
+
+		select {
+		case <-periodic.Next():
+		case <-process.GracefulExitContext().Done():
+			return nil
+		}
+	}
+}
+
+func (process *TeleportProcess) updateDeployServiceAgents(ctx context.Context, authClient *auth.Client) error {
+	cmc, err := authClient.GetClusterMaintenanceConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var criticalEndpoint string
+	if automaticupgrades.GetChannel() != "" {
+		criticalEndpoint, err = url.JoinPath(automaticupgrades.GetChannel(), "critical")
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	critical, err := automaticupgrades.Critical(process.GracefulExitContext(), criticalEndpoint)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !withinUpgradeWindow(cmc, process.Clock) && !critical {
+		return nil
+	}
+
+	teleportVersion, err := getStableTeleportVersion(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	issuer, err := awsoidc.IssuerFromPublicAddress(process.proxyPublicAddr().Addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	clusterNameConfig, err := authClient.GetClusterName()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	clusterName := clusterNameConfig.GetClusterName()
+
+	databases, err := authClient.GetDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
```

Comment:
This function needs comments explaining key bits of business logic in relevant parts of the function. For example, how we're collecting regions, that we're using OIDC integration clients, etc.

I know it's technically all in the code but it may be confusing for new readers and comments really help with code comprehension.